url8af84d296f_mrcosta_gantt_generator_tgz-pJ8-org_gantt_generator_mrcpsp_RunJ8/src/org/swiftgantt/model/Task.java

```diff
@@ -0,0 +1,622 @@
+					//Actual task end time as per start time.
+					this.actualEnd = TaskHelper.increaseCloneTimeByTimeUnit(tu, this.end, offset);
+				} else {
+					this.actualStart = this.start.clone();
+					this.actualEnd = this.end.clone();
+				}
+			}
+			if (logger.isDebugEnabled()) {
+				logger.debug("    After adjustment: " + this);
+			}
+		}
+		//Recursive to sub-tasks.
+		for (Task st : this.getChildren()) {
+			Task p = st.getLatestPredecessor();
+			st.adjustActualTimesByPredecessor(p);
+		}
+	}
+
+	/**
+	 * Get actual start time of this task.
+	 * 
+	 * @return
+	 */
+	public Time getActualStart() {
+		return this.checkAndLimitStartTime(actualStart);
+	}
+
+	/**
+	 * Set actual start time of this Task.
+	 *
+	 * @param actualStart
+	 */
+	protected void setActualStart(Calendar actualStart) {
+		this.setActualStart(new Time(actualStart));
+	}
+
+	/**
+	 * Set actual start time of this Task.
+	 * 
+	 * @param actualStart
+	 */
+	protected void setActualStart(Time actualStart) {
+		this.actualStart = actualStart;
+	}
+
+	/**
+	 * Get end time of this <code>Task</code>.
+	 * 
+	 * @return the end
+	 * @since 0.4.0
+	 */
+	public Time getEnd() {
+		Time endTime = this.checkAndLimitEndTime(this.end);
+		//		endTime.add(Time.DATE, 1);//Temprarily, will be change at v0.3
+		return endTime;
+	}
+
+	/**
+	 * Set end time of this <code>Task</code>.
+	 * @param end
+	 */
+	public void setEnd(Calendar end) {
+		this.setEnd(new Time(end));
+	}
+
+	/**
+	 * Set end time of this <code>Task</code>.
+	 * 
+	 * @param end The end time of task to set
+	 * @since 0.4.0
+	 */
+	public void setEnd(Time end) {
+		if (end == null) {
+			return;
+		}
+		Object oldValue = this.end;
+		this.end = end;// Clone to not to change the start time for outside.
+		//Adjust end time by latest predecessor task
+		if (this.start != null && this.predecessors.size() > 0) {
+			Task p = super.getLatestPredecessor();
+			if (p != null && p.getEnd() != null && p.getEnd().after(start) || p.getEnd().equals(start)) {
+				adjustActualTimesByPredecessor(p);
+			} else {
+				this.actualEnd = end.clone();
+			}
+		} else {
+			this.actualEnd = end.clone();
+		}
+		//also, as predecessor, set end time will affect sequence tasks.
+		for (Task t : this.subsequences) {
+			t.adjustActualTimesByPredecessor(this);
+		}
+		this.firePropertyChanged("End Time", oldValue, end);
+	}
+
+	/**
+	 * Get actual end time of this task.
+	 * 
+	 * @return
+	 */
+	public Time getActualEnd() {
+		return this.checkAndLimitEndTime(actualEnd);
+	}
+	/**
+	 * Set actual end time of this Task.
+	 *
+	 * @param actualEnd
+	 */
+	protected void setActualEnd(Calendar actualEnd) {
+		this.setActualEnd(new Time(actualEnd));
+	}
+
+	/**
+	 * Set actual end time of this Task.
+	 * 
+	 * @param actualEnd
+	 */
+	protected void setActualEnd(Time actualEnd) {
+		this.actualEnd = actualEnd;
+	}
+
+	/*
+	 * Calculate duration by time unit, start time and end time.
+	 */
+	private int calcDuration() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.